### ISSUE TITLE
ValidatePreconditions always fails if the supplied DateTimeOffset has non-zero milliseconds

### ISSUE THREAD

**Issue by cmeeren** [giraffe-fsharp/Giraffe#424]
Consider the following `DateTimeOffset`:

```f#
DateTimeOffset(2000, 1, 1, 0, 0, 0, TimeSpan.Zero)
```

This matches the following `If-Unmodified-Since` value (obtained by calling `.ToString("R")` on the `DateTimeOffset`):

```f#
"Sat, 01 Jan 2000 00:00:00 GMT"
```

And if I pass that `DateTimeOffset` to `HttpContext.ValidatePreconditions`, then that header value shown above will validate successfully.

However, most `DateTimeOffset` values are not exact seconds. If I add between 1 and 999 milliseconds

```f#
DateTimeOffset(2000, 1, 1, 0, 0, 0, 1, TimeSpan.Zero)
DateTimeOffset(2000, 1, 1, 0, 0, 0, 999, TimeSpan.Zero)
```

then, even though `.ToString("R")` returns the exact same string, the validation will fail, and AFAIK **there will be no way of making it pass**, because the resolution of the `If-Unmodified-Since` header is seconds.

This can be worked around in user code by manually rounding the `DateTimeOffset` down to the nearest second before passing the value to `ValidatePreconditions`, but ideally this should be fixed in Giraffe.

**Comment by cmeeren** [comment]
Could anyone confirm whether this is a bug in Giraffe? It's blocking a lot of work for us, so I need to know whether I should temporarily fix this in my own framework [Felicity](https://github.com/cmeeren/Felicity) (which builds on Giraffe), or perhaps I have misunderstood the HTTP spec.

**Comment by TheAngryByrd** [comment]
Since we're using DateTimeOffset directly in 

https://github.com/giraffe-fsharp/Giraffe/blob/612aa186c68101f532196817f67ebaa083c16146/src/Giraffe/Preconditional.fs#L61-L62 and https://github.com/giraffe-fsharp/Giraffe/blob/612aa186c68101f532196817f67ebaa083c16146/src/Giraffe/Preconditional.fs#L91-L92

This check will fail.  We'd have to go through the process of zeroing out the millisecond field.

In the mean time to unblock you, can you normalize the DateTimeOffset you're passing in to zero out the milliseconds?

**Comment by TheAngryByrd** [comment]
@dustinmoris what are you thoughts about fixing this? Since the [format](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/If-Unmodified-Since) doesn't seem to include milliseconds and [as specified in this RFC](https://www.w3.org/Protocols/rfc2616/rfc2616-sec3.html#sec3.3.1), should we go about zeroing out the milliseconds for the consumer?

**Comment by cmeeren** [comment]
Also see [this SO answer](https://stackoverflow.com/a/18603265/2978652) which also says milliseconds is invalid.

**Comment by cmeeren** [comment]
Note that I tried working around it using

```f#
dt.AddMilliseconds (float (-dt.Millisecond))
```

but that doesn't work in all cases, I guess due to rounding errors. The following seems to work, however:

```f#
dt.AddTicks(-(dt.Ticks % TimeSpan.TicksPerSecond))
```

**Comment by dustinmoris** [comment]
> should we go about zeroing out the milliseconds for the consumer?

Just by quickly reading through this thread and some links posted here I think that's what we should do.

**Comment by dustinmoris** [comment]
I just created a PR for the two of you to review. I might have misunderstood this so please feel free to correct me if I've got it wrong!

**Comment by dustinmoris** [comment]
Fixed with `5.0.0-rc-2`.
